Add border support to site logo

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Add border block support to the site logo.
Partial for https://github.com/WordPress/gutenberg/issues/43247, https://github.com/WordPress/gutenberg/issues/37757

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
To allow more customization of the block, and for consistency with other blocks.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The border is applied to the image element, not the wrapper.

## Testing Instructions
Please test both the placeholder and the site logo with image, with and without link, duotone, rounded style variation, and padding.
When testing with dutotone, you should find that the duotone changes the border color. This is consistent with other image blocks that supports both.
When testing padding, you should find that the padding is on the wrapper, which means it is not between the image and the border but between the border and other page content. This is also consistent with other blocks.

1. Activate a block theme that does not style the logo, for example, Twenty Twenty-Three.
2. Open the Site Editor.
3. Add a site logo block.
4. Next, open the Styles sidebar. Open the blocks panel and select the site logo.
5. Confirm that there is  a new button for opening the border styles panel. Open the panel and adjust the border.
If you are using a placeholder, you will see that the border has a low opacity, but it should be present.
I found that this is inconsistent across blocks. The image block placeholder has no visible border, the featured image has a fully visible border, not a border with low opacity.
Confirm that the border displays correctly when an image is selected.
6. Save and view the front. Confirm that the logo and border shows correctly.
7. Return to the site editor. With the block selected, open the block settings sidebar and locate the border panel.
8. Adjust the border. Please test both palette color and custom border color. The settings should override the border from the global styles.


## Screenshots or screencast <!-- if applicable -->

<img width="1507" alt="global styles border option for site logo" src="https://user-images.githubusercontent.com/7422055/221100803-ca97bc6b-fd0b-43b4-91d9-a530f34fd984.png">

carolinan committed:
Add border support to site logo

carolinan committed:
Update index.php

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

mikachan reviewed (approved):
Thank you for working on this! I've tested it out, and in my testing, it's working really well.

- Confirm that there is a new button for opening the border styles panel. ✅
- Confirm that the border displays correctly when an image is selected. ✅
- Confirm that the logo and border shows correctly. ✅
- The settings override the border from the global styles. ✅
- Tested with various custom border colours, styles, and surrounding global styles like padding and margin. ✅

aaronrobertshaw commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -122,7 +126,7 @@ const SiteLogo = ( {
 	const img = (
 		<img
-			className="custom-logo"
+			className={ `custom-logo ${ borderProps.className }` }
```
We should use the `classnames` util here and handle when `borderProps.className` is undefined. Otherwise, we'll end up with `custom-logo undefined` as the class name.

aaronrobertshaw reviewed (commented):
Thank you for taking the time to put this PR together @carolinan 👍 

There are a few issues we'll need to address before this can be merged.

1. The block needs to take border width into account with its sizing
2. The inline cropping should also reflect the border
3. The image class name can get `undefined` within it


https://github.com/WordPress/gutenberg/assets/60436221/680546e5-92d2-48ca-bb76-4b844d77d508

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Add a check for the border CSS classes

carolinan committed:
Add the custom border to the ImageEditor / cropper

carolinan commented:
Thank you. I did not realize the border needed to be included in the width; I expected it to be on the outside so that the image doesn't shrink.

> The block needs to take border width into account with its sizing
> The inline cropping should also reflect the border

I used the image block as a reference but did not want to copy code without fully understanding it.

**Width:** I added the `box-sizing: border-box;` to the image.
Problem: the right side resize handle is displayed correctly, but the bottom handle is not in the correct position, so I assume I need to adjust something related to the height.
However, when the block is centered, all resize handles are displayed correctly.

<img width="1260" alt="The bottom resize handle for the site logo block is in the wrong position" src="https://github.com/WordPress/gutenberg/assets/7422055/2bd742ee-da2e-43a7-82ae-f67f75f4a9f8">




**Crop:** I added the border to the cropper. I don't know this feature well enough to test if it works correctly now.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

aaronrobertshaw reviewed (commented):
Appreciate the effort and patience iterating on this one. I know there are a lot of edge cases around borders that can make these things a bit of a grind. 

> Problem: the right side resize handle is displayed correctly, but the bottom handle is not in the correct position, so I assume I need to adjust something related to the height.

The resizable box doesn't appear to be getting a height that matches the image. I wonder if tweaking the `lockAspectRatio` prop the same as the image block does would help?

> However, when the block is centered, all resize handles are displayed correctly.

Looking closer, it appears that the difference between the centred block and one without alignment is that the resizable box element gets a `display: table` style. Blindly adding that in via dev tools appears to put the resizable box's bottom handle where it should be.

I haven't gotten to explore why this is the case or what side effects might occur as a result but maybe that is a thread to follow 🤷 

> Crop: I added the border to the cropper. I don't know this feature well enough to test if it works correctly now.

You should be able to test the cropper by comparing it against trunk. Adding the border shouldn't change much other than the window through which the cropper operates. On that container, we might need to add the box-sizing rule to it as well. 

These croppers are due for some love in Phase 3 as part of the Media Library improvements so maybe a holistic solution might come through that.

One last thought, it was requested that the image block's placeholder not show the border when selected. Perhaps we should be doing the same here for the Site Logo? This would avoid the upload icon getting clipped at the edges as well.

carolinan commented:
@aaronrobertshaw Thank you, I will continue on it this week or next.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
The site logo block does not have aspect ratio controls, so I am unsure what parts to reuse from the image block.

With 6.3:
1) When the logo block is centered, the bottom control is no longer in the correct position. Now the bottom control is always misplaced.
2) I cannot get the ` display: table` trick to work.
3) Setting the `ResizableBox` height in edit.js to `auto` works. I am unsure if there are any problems with setting it auto, or what potential issues to test for.
```
			<ResizableBox
				size={ {
					width: currentWidth,
					height: 'auto',
				} }
```

carolinan commented:
Found another bug, but I see it on the image block too: after using the cropping tool, zooming, rotating, the horizontal resize handle is in the wrong position.

aaronrobertshaw commented:
> When the logo block is centered, the bottom control is no longer in the correct position. Now the bottom control is always misplaced.

I can't replicate this issue however while testing I did find that in the editor the styling for the Site Logo seems incorrect when it is set to link to home. 

In that case, the `a` and `img` aren't being set to fill the resizable box's dimensions. So with this styling issue, if the site logo block is drag resized beyond the dimensions of the image used for the Site Logo, it will appear as though the handles are way off.

https://github.com/WordPress/gutenberg/assets/60436221/b7b1cff2-5ae6-4d56-9628-6eaeda1bc497


> I cannot get the  display: table trick to work.

Using this updated branch and either 6.2 or 6.3-RC4, the handles look fine (except issue noted above), so we can probably forget this.

> Found another bug, but I see it on the image block too: after using the cropping tool, zooming, rotating, the horizontal resize handle is in the wrong position.

The only way I found the handles to be incorrectly positioned here was when the block was set to link to home again. 



Perhaps I'm missing something? Can you provide some replication steps or a video?

carolinan committed:
try height: auto, remove border when placeholder is selected

carolinan commented:
Image block resize handler test:
This time I did not use rotate, but I used the aspect ratio. It looks like it is not updated until the handle is selected.

https://github.com/WordPress/gutenberg/assets/7422055/8e370253-5457-488a-ae7a-75463b8ef72d

carolinan commented:
The default border color in the cropper is not correct:
https://github.com/WordPress/gutenberg/assets/7422055/ee28135a-e866-4d49-b6be-a51ca998b570

If there are resize handler issues with the aspect ratio, it seems to only be for some ratios. No issues here:
https://github.com/WordPress/gutenberg/assets/7422055/3d2e99da-2d17-4c0b-bc8b-56cdc3779913

No issues found with rotate:
https://github.com/WordPress/gutenberg/assets/7422055/4ea6942b-2fc2-4f46-9aec-5e7fc1d046e2

carolinan committed:
try to fix linked image width

carolinan commented:
With 100% width on the link and the image in the resizer, the logo is much smaller after you click on _apply_, so I don't think this is correct:

https://github.com/WordPress/gutenberg/assets/7422055/c4ebdeb5-fbe4-4e33-bb2c-e90e5185dc13


Edit: It might work better if you put the CSS in the single `.wp-block-site-logo `selector, not the placeholder CSS...

carolinan committed:
Move the CSS for the linked image width

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

aaronrobertshaw commented:
Thanks for continuing to dig into the nuances of the cropper etc. It sounds like we'll need a dedicated issue detailing all the edge cases to fix there.

> This time I did not use rotate, but I used the aspect ratio. It looks like it is not updated until the handle is selected.

This might just have been an oversight when the aspect ratio functionality was added/updated, that it didn't force the resizable box to update its dimensions as well as the image.

> With 100% width on the link and the image in the resizer, the logo is much smaller after you click on apply, so I don't think this is correct:

It's a shame that a simple `width: 100%` won't fix the issue. The linked site logo still needs fixing one way or another. It might still be best to address that while also adding borders to ensure that any solution isn't impacted negatively by the that addition.

carolinan commented:
hmm, looking at that video again, it looks like the image becomes larger when the crop/zoom, etc tools are opened, and then it returns to the original size after applying.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
CS: Rename function name

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
This is an image block with a border:
![image](https://github.com/WordPress/gutenberg/assets/7422055/f35a7bb5-5981-4ab9-a0e8-3878cc47d6b0)
The same image, with the cropping tool opened:
![image](https://github.com/WordPress/gutenberg/assets/7422055/c4b07de6-7223-4c90-9385-aa65ded31ec0)

So the size change when the cropping tool opens is not a bug with the site logo.
But the background color is different.

Bordered Site logo without the cropping tool opened:
<img width="715" alt="Site logo block with orange border" src="https://github.com/WordPress/gutenberg/assets/7422055/dcdff3e6-f51d-416d-bd43-66e1bc326bda">

Bordered Site logo with the cropping tool opened:
<img width="715" alt="Site logo block with orange border and the cropping tool opened" src="https://github.com/WordPress/gutenberg/assets/7422055/f0dcee1b-f76c-4591-a34c-bcd4675fcb12">

carolinan commented:
This gets weirder: When the image rotate option is used, the background color is the same for both the image block and the site logo block, but the size is different.

Both blocks are set to be 250x250px.

Image:
<img width="683" alt="Image block with a border and the image rotation option opend" src="https://github.com/WordPress/gutenberg/assets/7422055/c45144f7-52fc-4f64-9734-f34bc7ac9c7a">

Site Logo, barely fits on my screen:
<img width="683" alt="Screenshot 2023-10-13 at 08 39 06" src="https://github.com/WordPress/gutenberg/assets/7422055/a7dd2230-cc18-4782-a252-37268f659409">

carolinan commented:
I am not sure how that height can be fixed without adding the height to the site logo. It currently only has a width setting.

carolinan commented:
This is the cropping tool container, and without knowing the code well enough yet, I would say that the height here is actually the content width 🤔  ( I am testing in Twenty Twenty-Four)

`<div class="wp-block-image__crop-area has-border-color" style="border-color: rgb(216, 97, 60); border-width: 20px; border-radius: 10px; width: 250px; height: 620px;">
`

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Fix editing tool border color by moving the <a> element

carolinan commented:
Update: To fix the default border color when the image editing tools in the block toolbar are used, I needed to move the `<a>` element so that it was _outside_  the crop area and the resizable container.
I reused code from the image block:  https://github.com/WordPress/gutenberg/pull/55470

aristath commented:
@carolinan Has this been fixed? I believe the issue discussed above has already been resolved in the other PR? If so, we can merge this 👍

t-hamano reviewed (changes requested):
Sorry, I've found some bugs in this PR, so I'd like to pause it.

t-hamano commented:
- When I click on the block, I don't see the blue outline. It's probably related to the fact that [the click is disabled](https://github.com/WordPress/gutenberg/pull/48354/files#diff-0d5debfeb0c6fe28a3979c43ecb9029bae1617d404b1c92f506ba754333f9647R399-R402). This may be due to this, but for example, the following keyboard operations do not work.
  - Enter key to insert next block
  - Delete key to delete block
- Turning off the "Link image to home" option will break the block.

https://github.com/WordPress/gutenberg/assets/54422211/68e36c89-9e93-4e83-99db-bd645297becd

t-hamano commented:
> When I click on the block, I don't see the blue outline. It's probably related to the fact that [the click is disabled](https://github.com/WordPress/gutenberg/pull/48354/files#diff-0d5debfeb0c6fe28a3979c43ecb9029bae1617d404b1c92f506ba754333f9647R399-R402). This may be due to this, but for example, the following keyboard operations do not work.

For this problem, applying `pointer-events: none;` to the `a` element may be effective. See #56043

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Fix the JavaScript error when the link option is toggled off

carolinan commented:
I am trying a variation of the changes in 56043 but running into a problem where the logo is not resizable with the drag handles when the link is enabled.

t-hamano commented:
> the logo is not resizable with the drag handles when the link is enabled.

I think this is because the `ResiableBox` is wrapped in an `a` element with `pointer-events: none;` applied. This can be resolved by moving the `a` element into the `Resizablebox`. (See [this comment](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1810356718))

On the other hand, another problem has occurred due to moving the `a` element. (See [this comment](https://github.com/WordPress/gutenberg/pull/56043#issuecomment-1814447328))

I'm currently searching for a solution to this problem.

t-hamano commented:
The approach taken in #56043 appears to have worked ideally in the end. That code may be helpful for this PR.

carolinan committed:
Update the <a> image wrapper and editor styles. Add 'has-custom-boder' class name to index.php

carolinan commented:
This is ready for another review. 
There is one inconsistency left that I am aware of:
When the image block has a border and the rounded style, then opening the crop tool in the toolbar hides the border.
When you do the same with the site logo, the border is visible.

aaronrobertshaw reviewed (commented):
I ran into a couple of issues while testing the latest version of this PR.

When the site logo is set to "Link image to home":
- Choosing the crop tool from the block toolbar causes a centre aligned site logo to shift suddenly as if it has no alignment.
- You cannot scroll with the mouse wheel to zoom in or out on the image within the crop tool.

When not wrapping the logo image in a link, both of the above features still work.

| Trunk | This PR |
|---|---|
| <video src="https://github.com/WordPress/gutenberg/assets/60436221/a55920a4-82b0-4d85-b5e4-bb2aa2fa301a" /> | <video src="https://github.com/WordPress/gutenberg/assets/60436221/e5ac36a3-5926-4ec9-bd9f-78275b17ddeb" /> |


> When the image block has a border and the rounded style, then opening the crop tool in the toolbar hides the border.

I think this is outside the scope for this PR and could be addressed in a follow-up. There's a check in the image block's code that looks for the rounded block style and if present does not pass the border props through to the image editor.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Add display:table to centered logos

carolinan commented:
@ramonjd @t-hamano If possible I am going to need your combined brainpower because the `pointerEvents: 'none'` style that was added to both the image block, and to the site logo block in this PR, is breaking the mouse scroll zoom in the cropping tool, as @aaronrobertshaw first found in testing.

If I comment out the `pointerEvents: 'none'` style the zoom works again.

https://github.com/WordPress/gutenberg/pull/56043/files#diff-a14b61c874095649b236b78413e805b4541277b24eb71ce350a17d9f2b6396b1R101

t-hamano commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -58,6 +76,59 @@ function render_block_core_site_logo( $attributes ) {
+
+	// Border color.
+	$preset_color           = array_key_exists( 'borderColor', $attributes ) ? "var:preset|color|{$attributes['borderColor']}" : null;
+	$custom_color           = _wp_array_get( $attributes, array( 'style', 'border', 'color' ), null );
```
```suggestion
	$custom_color           = $attributes['style']['border']['color'] ?? null;
```

From a performance standpoint, `_wp_array_get()` function may be best avoided. See #51116

t-hamano commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -58,6 +76,59 @@ function render_block_core_site_logo( $attributes ) {
+
+	// Individual border styles e.g. top, left etc.
+	foreach ( $sides as $side ) {
+		$border                 = _wp_array_get( $attributes, array( 'style', 'border', $side ), null );
```
```suggestion
		$border                 = $attributes['style']['border'][ $side ] ?? null;
```

t-hamano commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -34,6 +34,8 @@ function render_block_core_site_logo( $attributes ) {
 	if ( ! $attributes['isLink'] ) {
 		// Remove the link.
 		$custom_logo = preg_replace( '#<a.*?>(.*?)</a>#i', '\1', $custom_logo );
+
+		preg_replace( '#<a.*?>(.*?)</a>#i', '\1', $custom_logo );
```
I think this line is unnecessary because there is the same code just before.

carolinan replied:
I don't remember adding this, must have been a copy-paste typo 😅. Thank you.

t-hamano commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -162,7 +177,7 @@ const SiteLogo = ( {
 
 	if ( ! isResizable || ! imageWidthWithinContainer ) {
-		return <div style={ { width, height } }>{ imgWrapper }</div>;
+		return <div style={ { width, height } }>{ img }</div>;
```
```suggestion
		return (
			<div style={ { width, height } }>
				<ImageWrapper
					isLink={ isLink }
					href={ siteUrl }
					title={ title }
					classes={ classes }
				>
					{ img }
				</ImageWrapper>
			</div>
		);
```

In the original code, the `img` element should have been wrapped in an `a` element if `isLink` was `true`.

t-hamano reviewed (commented):
Thank you for tackling this difficult issue, @carolinan!

> If I comment out the `pointerEvents: 'none'` style the zoom works again.

The `pointer-events` property should have been overridable from child elements. I quickly tried adding `pointer-events:auto` with `.wp-block-image__crop-area` as the selector, and zooming and mouse dragging seem to work.

ramonjd commented:
> tried adding pointer-events:auto with .wp-block-image__crop-area as the selector, and zooming and mouse dragging seem to work.

Can confirm that the follow fixes it up. Thanks @t-hamano !

```diff
diff --git a/packages/block-library/src/site-logo/editor.scss b/packages/block-library/src/site-logo/editor.scss
index b219216782..ed03c53072 100644
--- a/packages/block-library/src/site-logo/editor.scss
+++ b/packages/block-library/src/site-logo/editor.scss
@@ -24,6 +24,9 @@
 		height: auto;
 		max-width: 100%;
 	}
+	.wp-block-image__crop-area {
+		pointer-events: auto;
+	}
 }
 
 // Provide special styling for the placeholder.

```

ramonjd commented on packages/block-library/src/site-logo/style.scss:
```diff
@@ -20,6 +20,12 @@
 		max-width: 100%;
 	}
 
+	&.has-custom-border {
+		img {
+			box-sizing: border-box;
+		}
+	}
```
```suggestion
	&.has-custom-border img {
		box-sizing: border-box;
	}
```

optional, but it's a little less work for the SASS compiler :)

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Code cleanup based on review

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

t-hamano commented:
Sorry for the late reply.

I tested various scenarios in this PR to see what was expected and if there were any remaining issues that were discovered during the discussion that needed to be resolved.

- ✅ When the logo is linked, inherit the link color
- ✅ If no logo is registered, i.e. when a placeholder is being rendered, the border will not be rendered (this is expected behavior, right?)
- ✅ Border thickness has no effect on image width, whether the logo is linked or not
- ✅ Regardless of whether the logo is linked or not, and regardless of the left or right position of the image, the resizing handle will not shift from the image.
- ✅ Enabling the crop tool with a centered logo does not shift the logo to the left.
- ✅ When the crop tool is enabled, the default border color is inherited when the logo is not linked, and the link color is inherited when the logo is linked.
- ✅ In the crop tool, zooming and dragging using the mouse works.
- ✅ When rotating an image with the crop tool, the height does not seem to become extremely high.

Below are what I currently think are problems that may need to be solved. If there are any other issues that need to be resolved, please let me know.

- When the logo image is small and linked, the image does not fill the width of the resizable box (this also seems to happen with trunk).
- The image becomes slightly larger when the crop tool is enabled (This also seems to be happening with image blocks).
- When a border is applied, the height of the image changes slightly when changing the link state.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
> The image becomes slightly larger when the crop tool is enabled (This also seems to be happening with image blocks).

I am able to solve this by removing an outer border from the crop area ( .reactEasyCrop_CropArea). But when I do that, the grey background of the cropping tool is visible on the right and left side of the image.

In the example, the logo has a 2px solid red border.

With the cropping tool closed:
![Site logo block with the cropping tool closed](https://github.com/WordPress/gutenberg/assets/7422055/356e6df5-8a51-44fc-ae03-180f7498b1c4)

With the cropping tool opened:
![Site logo block with the cropping tool opened](https://github.com/WordPress/gutenberg/assets/7422055/e80a35f9-4fec-43a1-bd2d-5a4e6a284fa6)

carolinan committed:
Update editor.scss

carolinan commented:
> When a border is applied, the height of the image changes slightly when changing the link state.

The image had the height set to `inherit`, so the height changed when the parent element changed between the `<a>` and the `<div>`.

Removing it and falling back to `auto` seems to solve it, but I am not sure if there are any side effects that I have not considered.

carolinan committed:
Try to fix the height difference that occurs when the parent element changes

t-hamano commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -122,7 +154,9 @@ const SiteLogo = ( {
 	const img = (
 		<img
-			className="custom-logo"
+			className={ classnames( 'custom-logo', {
+				[ borderProps.className ]: !! borderProps.className,
+			} ) }
```
```suggestion
			className={ classnames( 'custom-logo', borderProps.className ) }
```

I think the undefined check is unnecessary.

t-hamano commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -28,6 +24,10 @@
 		max-width: 100%;
 	}
+
+	.wp-block-image__crop-area {
+		pointer-events: auto;
+	}
```
Note: As a follow-up, we'll want to add this style to the Image block as well. The current Image block does not accept mouse operations when the Crop tool is enabled.

t-hamano commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -151,3 +151,25 @@
+}
+
+.wp-block-site-logo.has-custom-border {
+	.reactEasyCrop_CropArea {
+		border-width: 0;
+	}
```
> > The image becomes slightly larger when the crop tool is enabled (This also seems to be happening with image blocks).
> 
> I am able to solve this by removing an outer border from the crop area ( .reactEasyCrop_CropArea). But when I do that, the grey background of the cropping tool is visible on the right and left side of the image.

Is it possible to delete this style?

The problem of the image becoming slightly larger when the crop tool is enabled seems to be resolved simply by the following style, which has already been added in this PR.

```scss
.wp-block-image__crop-area {
	box-sizing: border-box;
}
```

carolinan replied:
I didn't like the result of removing the border width, but without it I still do see the size difference.

carolinan replied:
I also don't think the blocks should override these type of styles; the crop tool itself may need updates.

t-hamano reviewed (commented):
Thank you, for your continued work on this issue, @carolinan! My feeling is that it's almost complete, but I'd like to make a few additional comments.

t-hamano commented:
Nit; Although not added in this PR, the following styles also exist in style.scss, so they can be deleted.

https://github.com/WordPress/gutenberg/blob/2e987e5add9c7ddc32e69e43ddecff4b729b1d5b/packages/block-library/src/site-logo/editor.scss#L24-L25

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Remove undefined check for the borderProps.className

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

carolinan committed:
Remove unwanted and duplicate CSS in editor.scss

t-hamano reviewed (approved):
LGTM! From what I've tested, I think all the bugs reported so far (related to this PR) have been fixed, but I'd like to hear @aaronrobertshaw's review again 🙇

Also, through this PR, we were able to discover several bugs, including those related to the Image block. These are also occurring on trunk, so I think they can be addressed with follow-up PRs. As far as I've researched, I think the following issues need to be fixed.

1. When rotating the Site Logo block image, there is an unnatural space at the top and bottom:

![image](https://github.com/WordPress/gutenberg/assets/54422211/8c3d0685-0fa3-4625-88fa-8c3c0911be8b)

2. The crop tool cannot be operated with the mouse when the Image block is linked. It is necessary to apply `pointer-events: auto;` to the crop area.

![image](https://github.com/WordPress/gutenberg/assets/54422211/b9c28ed8-7d4d-4b0e-bd88-945ee17ebcce)

3. Move the CSS related to the crop tool (`ImageEditor` component) that is duplicated between the Site Logo block and Image block to the component itself

aaronrobertshaw commented on packages/block-library/src/site-logo/block.json:
```diff
@@ -33,6 +33,20 @@
+			"radius": true,
+			"width": true,
+			"style": true,
+			"__experimentalSelector": "img, .components-placeholder__illustration, .components-placeholder::before, .wp-block-image__crop-area",
```
This should use the [block selectors API](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-selectors.md) rather than the outdated `experimentalSelector` property, as per the [image block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/block.json#L122-L127).

carolinan replied:
If I try to move both the duotone and the border to selectors, the duotone changes the color of the resize handles. I am not sure how that is prevented in the image block.

carolinan replied:
I'm getting confused, the more I test this, the more it looks like the selector for the border is not even needed.

t-hamano replied:
Does the change below not work? Personally, I have submitted #58035 regarding the deprecated duotone selector, so I think it can be removed from the scope of this PR.

```diff
diff --git a/packages/block-library/src/site-logo/block.json b/packages/block-library/src/site-logo/block.json
index 977a042a72..44c08e60da 100644
--- a/packages/block-library/src/site-logo/block.json
+++ b/packages/block-library/src/site-logo/block.json
@@ -38,7 +38,6 @@
                        "radius": true,
                        "width": true,
                        "style": true,
-                       "__experimentalSelector": "img, .components-placeholder__illustration, .components-placeholder::before, .wp-block-image__crop-area",
                        "__experimentalSkipSerialization": true,
                        "__experimentalDefaultControls": {
                                "color": true,
@@ -48,7 +47,6 @@
                        }
                },
                "color": {
-                       "__experimentalDuotone": "img, .components-placeholder__illustration, .components-placeholder::before",
                        "text": false,
                        "background": false
                },
@@ -59,6 +57,15 @@
                                "margin": false,
                                "padding": false
                        }
+               },
+               "filter": {
+                       "duotone": true
+               }
+       },
+       "selectors": {
+               "border": ".wp-block-site-logo img, .components-placeholder__illustration, .wp-block-site-logo .components-placeholder::before, .wp-block-site-logo .wp-b
lock-image__crop-area",
+               "filter": {
+                       "duotone": ".wp-block-site-logo img, .wp-block-site-logo .components-placeholder__illustration, .wp-block-site-logo .components-placeholder::befo
re"
                }
        },
        "styles": [
```

carolinan replied:
I had set 	color :` "__experimentalDuotone": true`, that's what was affecting the resize handlers.

carolinan replied:
But if I just delete  __experimentalBorder:
`"__experimentalSelector": "img, .components-placeholder__illustration, .components-placeholder::before, .wp-block-image__crop-area",`
I don't see anything breaking.

t-hamano replied:
> I don't see anything breaking.

If you simply remove `__experimentalBorder.__experimentalSelector`, the border functionality at the block level may still work. However, when applying a border to the Site Logo block via the Global Styles, the border will not be applied to the intended element. My understanding is that this property solves this, and I think it's the intended behavior.

carolinan replied:
OK I think I'm having one of those days where I shouldn't be close to a computer 😁

aaronrobertshaw commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -58,6 +74,59 @@ function render_block_core_site_logo( $attributes ) {
+ * @param array $attributes The block attributes.
+ * @return array The border-related classnames and styles for the block.
+ */
+function block_core_site_logo_get_border_attributes( $attributes ) {
```
There's a bit of inconsistent use of the null coalescing operator in this function I think we could lean into using it which would make the code a bit easier to grok.

<details>
<summary>Example</summary>

```
diff --git a/packages/block-library/src/site-logo/index.php b/packages/block-library/src/site-logo/index.php
index 98891c3f5e..a12e0e2bf3 100644
--- a/packages/block-library/src/site-logo/index.php
+++ b/packages/block-library/src/site-logo/index.php
@@ -82,37 +82,25 @@ function render_block_core_site_logo( $attributes ) {
  * @return array The border-related classnames and styles for the block.
  */
 function block_core_site_logo_get_border_attributes( $attributes ) {
-
-	$border_styles = array();
 	$sides         = array( 'top', 'right', 'bottom', 'left' );
-
-	// Border radius.
-	if ( isset( $attributes['style']['border']['radius'] ) ) {
-		$border_styles['radius'] = $attributes['style']['border']['radius'];
-	}
-
-	// Border style.
-	if ( isset( $attributes['style']['border']['style'] ) ) {
-		$border_styles['style'] = $attributes['style']['border']['style'];
-	}
-
-	// Border width.
-	if ( isset( $attributes['style']['border']['width'] ) ) {
-		$border_styles['width'] = $attributes['style']['border']['width'];
-	}
+	$border_styles = array(
+		'radius' => $attributes['style']['border']['radius'] ?? null,
+		'style'  => $attributes['style']['border']['style'] ?? null,
+		'width'  => $attributes['style']['border']['width'] ?? null,
+	);
 
 	// Border color.
 	$preset_color           = array_key_exists( 'borderColor', $attributes ) ? "var:preset|color|{$attributes['borderColor']}" : null;
 	$custom_color           = $attributes['style']['border']['color'] ?? null;
-	$border_styles['color'] = $preset_color ? $preset_color : $custom_color;
+	$border_styles['color'] = $preset_color ?? $custom_color;
 
 	// Individual border styles e.g. top, left etc.
 	foreach ( $sides as $side ) {
 		$border                 = $attributes['style']['border'][ $side ] ?? null;
 		$border_styles[ $side ] = array(
-			'color' => isset( $border['color'] ) ? $border['color'] : null,
-			'style' => isset( $border['style'] ) ? $border['style'] : null,
-			'width' => isset( $border['width'] ) ? $border['width'] : null,
+			'color' => $border['color'] ?? null,
+			'style' => $border['style'] ?? null,
+			'width' => $border['width'] ?? null,
 		);
 	}
 

```
</details>

carolinan replied:
Similiar code is repeated in the featured image block and avatar, I wonder if it could be moved out of the blocks to lib > block supports > border

aaronrobertshaw reviewed (commented):
This is testing nicely for me,  thank you for all the hard work on it. 👍 

I left a couple of suggestions as inline comments. The main one to address is avoiding the `__experimentalSelector` property that we should be moving away from.

It would also be great if we can collect the bug fixes identified for follow-ups into new issues.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
And I found another problem, when the media is taller than it is wide, the bottom resize handle is not in the correct position, but too far down. This does not happen with the image block.

<img width="418" alt="The bottom resize handler on the selected site logo block has unexpected spacing between itself and the image border" src="https://github.com/WordPress/gutenberg/assets/7422055/47351b70-b76d-407f-910d-ac8872a8ff21">

carolinan committed:
Simplify the conditions and assignments in block_core_site_logo_get_border_attributes()

carolinan commented:
The resize handles in the Site Editor are not in the correct position:

https://github.com/WordPress/gutenberg/assets/7422055/a7deeef5-64ea-42fa-b629-66e4a8fd93f7

carolinan committed:
Move the border selector from supports to selectors.

t-hamano commented:
> And I found another problem, when the media is taller than it is wide, the bottom resize handle is not in the correct position, but too far down. This does not happen with the image block.

Does this issue also occur on trunk?

> The resize handles in the Site Editor are not in the correct position:
> (Note that one of the logos is inside the row, the other is outside, in the template part).

Maybe it has something to do with `box-sizing`. Is it possible to solve the problem by applying the `box-sizing: border-box` [here](https://github.com/WordPress/gutenberg/pull/48354/files#diff-e8d11e13318421cd345db3d9cd1e6104003bcb1dd044895fc4a57ce99f41696aR24) to the `img` element itself?

This is because the image block [also has this style](https://github.com/WordPress/gutenberg/blob/b8474c83aa37c7ab299b4f251a0f96ad598c5f95/packages/block-library/src/image/style.scss#L6) applied to the `img` element itself.

t-hamano reviewed (changes requested):
Block approval; This PR has been approved once, but there are still issues to be resolved

carolinan commented:
> > And I found another problem, when the media is taller than it is wide, the bottom resize handle is not in the correct position, but too far down. This does not happen with the image block.
> 
> Does this issue also occur on trunk?
> 
> > The resize handles in the Site Editor are not in the correct position:
> > (Note that one of the logos is inside the row, the other is outside, in the template part).
> 
> Maybe it has something to do with `box-sizing`. Is it possible to solve the problem by applying the `box-sizing: border-box` [here](https://github.com/WordPress/gutenberg/pull/48354/files#diff-e8d11e13318421cd345db3d9cd1e6104003bcb1dd044895fc4a57ce99f41696aR24) to the `img` element itself?
> 
> This is because the image block [also has this style](https://github.com/WordPress/gutenberg/blob/b8474c83aa37c7ab299b4f251a0f96ad598c5f95/packages/block-library/src/image/style.scss#L6) applied to the `img` element itself.

I have not found the time to test this yet.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Update style.scss

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Fix coding standard issues

Remove the default value for the boolean aria-disabled in edit.js
Add a since to block_core_site_logo_get_border_attributes() in index.php

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
I am no longer able to reproduce the issue in the Site Editor where the logo in the header would overlap the Site title block next to it, and where the horisontal resize handle was in the wrong position.

The bottom resize handle is still positioned a few pixels below the border. To answer the question, no it does not happen on trunk. With Gutenberg trunk active, the bottom resize handle is in the correct position.

carolinan commented:
Changing the image height from auto to 100% in the editor solves the position of the bottom resize handler in Chrome, does not fix it in Firefox, and completely hides (breaks) the image in Safari 😢

carolinan commented:
In addition, if the block has left or right margin, the center alignment option in the block toolbar stops working, with or without a border applied.
This does not happen on trunk.

carolinan committed:
Revert the change to .wp-block-site-logo.aligncenter.

Revert the change to .wp-block-site-logo.aligncenter, to fix a problem with the center alignment when the block has left or right margin set in the dimensions panel.

carolinan committed:
Try to fix merge conflicts, attempt 1

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
I think the small spacing between the bottom resize handler and the image is acceptable.
Current state of this PR:

Firefox, macOs

https://github.com/WordPress/gutenberg/assets/7422055/7ff7baf7-fe4d-4359-b3cb-d5a7eed250a2

Safari

https://github.com/WordPress/gutenberg/assets/7422055/cc13e586-89c6-4610-8c0f-e2833a39f19b

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Fix issues after the merge conflict

Adjust spacing.
replace classnames with clsx.

carolinan commented:
I found a new bug but it also happens on trunk and should not block this.
In the editors, the logo can not be resized larger than its original size, while on the front, it shows with the size that is set using the resize handlers or the image width range slider (and depending on the container).

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

aaronrobertshaw commented on packages/block-library/src/site-logo/block.json:
```diff
@@ -33,6 +33,19 @@
+			"width": true,
+			"style": true,
+			"__experimentalSkipSerialization": true,
+			"__experimentalDefaultControls": {
+				"color": true,
+				"radius": true,
+				"width": true,
+				"style": true
+			}
```
```suggestion
```

Recent feedback has been that if a block doesn't display its spacing controls by default then we should also make the border controls optional.

aaronrobertshaw commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -54,9 +55,37 @@ import { MIN_SIZE } from '../image/constants';
 const ACCEPT_MEDIA_STRING = 'image/*';
 
+// If the logo is linked, wrap in an <a /> tag to trigger any inherited link element styles.
+const ImageWrapper = ( { isLink, href, title, classes, children } ) => {
```
```suggestion
const ImageWrapper = ( { isLink, href, title, classes, children } ) => {
```

We should probably use the more standard `className` prop for this component.

carolinan replied:
I can't even remember what classes I was trying to add to the ImageWrapper and why.

aaronrobertshaw commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -67,12 +96,17 @@ const SiteLogo = ( {
 	const [ isEditingImage, setIsEditingImage ] = useState( false );
 	const { toggleSelection } = useDispatch( blockEditorStore );
+	const borderProps = useBorderProps( attributes );
+	const classes = clsx( 'custom-logo-link' );
```
Do we need the use of `clsx` here? Can't this just be passed as the value directly on the prop below?

carolinan replied:
This was already removed on trunk. 👍

aaronrobertshaw commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -51,6 +51,22 @@ function render_block_core_site_logo( $attributes ) {
+		$processor = new WP_HTML_Tag_Processor( $custom_logo );
+		$processor->next_tag( 'img' );
+		if ( ! empty( $border_attributes['class'] ) ) {
+			$processor->set_attribute( 'class', $border_attributes['class'] );
```
In case there are other classes already, should this use the processor's `add_class` method instead? Likewise could there be existing styles that need to be taken into account for the other `set_attribute` call?

carolinan replied:
Good catch, this code removed the "custom-logo" class from the img on the front.

aaronrobertshaw commented on packages/block-library/src/site-logo/index.php:
```diff
@@ -60,6 +76,49 @@ function render_block_core_site_logo( $attributes ) {
+ * Generates class names and styles to apply the border support styles for
+ * the site logo block.
+ *
+ * @since 6.6.0
```
Nit: This will need a bump in version.

aaronrobertshaw reviewed (commented):
I appreciate the persistence and hard work on this one @carolinan, I know it's a tricky one! 🙇 

After giving this one another run, some aspects are working well and others not so much.

In particular, we might need to improve the handling around placeholders. I'm getting some weird behaviour when the block is in its placeholder state, including:
- Scrollbars within the placeholder when borders are applied
    <img width="157" alt="Screenshot 2024-07-31 at 12 57 33 PM" src="https://github.com/user-attachments/assets/1b269741-556c-44dc-9980-ae4f8a7a9bcf">
- Somewhat inconsistent rendering of the placeholder, seemed warped when different border combinations were applied
- Global styles apply to the placeholder at all times but block instance styles don't apply to the placeholder when the block is selected, which it has to be to edit the styles. It's a bit confusing.

Also, the global border radius styles don't apply. The `.wp-block-site-logo img` style will need to have the selector's specificity lowered.

When the block instance styles are applied, the image cropper maintains its shape and size nicely. That isn't the case when only using global border styles or the rounded block style.

That might sound like a lot but this has come a long way. Nice work!


https://github.com/user-attachments/assets/b1021a6c-1f0a-47df-afd1-ed047fa5d638

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan committed:
Remove experimental default controls

carolinan committed:
Border attributes: Use add_class instead of set_attributes

carolinan committed:
Try to remove the CSS classes from the ImageWrapper to test if anything breaks.

carolinan committed:
Show block border styles when the placeholder is selected.

carolinan commented:
> Also, the global border radius styles don't apply. The .wp-block-site-logo img style will need to have the selector's specificity lowered.

On the placeholder, the CSS from the component is overriding the border radius from the global styles.
This also happens with the image block, on trunk.

carolinan committed:
WIP: update CSS reduce specificity

carolinan commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -47,24 +44,27 @@
 	&.is-default-size .components-placeholder {
 		height: 60px;
 		width: 60px;
 	}
```
If I don't remove the height and width, the upload button is misaligned when there is a border.
In this screenshot, a border width of 10px is added in global styles:
<img width="91" alt="logo-placeholder" src="https://github.com/user-attachments/assets/51da2c0d-079f-4d53-89f5-02e47831c668">

carolinan replied:
I am not sure what this CSS is for, was 60px chosen because the default size of the logo is 120?

carolinan replied:
Ok, I found the ticket https://github.com/WordPress/gutenberg/pull/48218
I am not sure how to resolve it without letting it fall back to `height: 100% `and `width:100%` though! @richtabor

richtabor replied:
> I am not sure what this CSS is for, was 60px chosen because the default size of the logo is 120?

Yes, per https://github.com/WordPress/gutenberg/pull/48218—120x120 is too big for a logo placeholder, but 120px is the max initial width for when a logo is uploaded.

aaronrobertshaw replied:
This is a tricky one 🤔 

While in the placeholder state the layout shift when adding a border sticks out. When not in the placeholder state the block has `border-box` set for the `box-sizing`. Adding `border-box` for the placeholder state exacerbates the problem with scrollbars etc.

I'm not sure there is a bulletproof solution but maybe there is something we can do to smooth out the experience a bit for the less extreme use cases.

If we avoid applying the custom border on the placeholder directly and instead apply it to a new pseudo element. We can position the bordered pseudo element over the inner contents. It will start to crop the underlying button as the border increases but that seems better than scrollbars.

Hacking around in dev tools this is what I got:


https://github.com/user-attachments/assets/18831190-918b-4b51-ac0e-98c0f484e21f

carolinan commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -47,24 +44,27 @@
 	> div, // A 60px width div shown only in the editor on mobile.
 	.components-resizable-box__container {
 		border-radius: inherit;
 	}
```
The `border-radius:inherit` on the `>div` here is overriding the border radius from the global styles.
I did not understand the code comment about the div only being shown in the editor on mobile, or how to test it.

carolinan commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -108,6 +108,12 @@
+:root :where(.wp-block-site-logo .components-placeholder) {
+	border-radius: inherit;
+}
+
```
Even with the `border-radius:inherit `
1) Commented out on line 58
2) Moved from line 67 to 114 with reduced specificity

the placeholder component's own CSS still overrides the border radius from the global styles.

carolinan commented on packages/block-library/src/site-logo/style.scss:
```diff
@@ -18,12 +18,7 @@
-	a,
-	img {
-		border-radius: inherit;
+		box-sizing: border-box;
```
I found that the border radius from the "rounded" style variation is applied without this CSS, but maybe I missed something in my testing.

richtabor commented:
I'm seeing this here; the default logo block experience should look like an image block, but be contained by a default width:

![CleanShot 2024-07-31 at 10 35 17](https://github.com/user-attachments/assets/93b07700-d0ab-4305-bf84-4a6c50e0599f)

Would it be fine to just omit the border on the placeholder? The radius is fine, but the border is a bit funky. I get its applied elsewhere; just looking for a way to unblock this.

carolinan commented:
I don't understand. What is "this* ? I don't know what setting is applied on the block in the screenshot or how it differs from the expected. Funky is not a technical term I can solve 😂😂😂

carolinan commented:
I should probably open a separate issue for this, but I want to ping @ciampo in this PR first because of the context.
I don't expect you to read all the comments on this PR :) The most recent comments are enough.

There seems to be conflicting intentions about wether the placeholder component should be styled by blocks or not.
Currently, the border radius in the placeholder component has a higher specificity than the border radius from global styles.
There is a comment in the SCSS file of the placeholder component that this is intentional:
```
// This needs specificity to override individual block styles.
.components-placeholder.components-placeholder {
```
https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/placeholder/style.scss#L1

So I am looking for a definite answer about wether the placeholder component should be styled by global styles or not.
The placeholders are already styled by the block settings since this CSS is inline, the question is about the global styles.
"Sometimes" may be a valid answer -But with the current CSS; I do not know how to resolve that.

carolinan committed:
Add the size limitation back

carolinan committed:
Try resetting the box-sizing on the placeholder.

carolinan commented:
For the part where the border on the placeholder is covering the misaligned upload button, what if I remove the box-sizing?
The logo placeholder will be larger than the 60x60, but at least the button is not cut off by the border.

https://github.com/user-attachments/assets/a4942ca4-b5b2-4c6e-b6dd-b6a002cd5dba

aaronrobertshaw commented:
Thanks for persisting on this one @carolinan, huge effort 💪 

> Currently, the border radius in the placeholder component has a higher specificity than the border radius from global styles.
> There is a comment in the SCSS file of the placeholder component that this is intentional:

I'd say at the time, this was intentional and desired. I also recall there was an effort to overhaul and improve placeholders more broadly. It's possible this is where things diverged. There might be a thread to follow in https://github.com/WordPress/gutenberg/pull/43180 and before that https://github.com/WordPress/gutenberg/pull/42847#issuecomment-1201030052.


> For the part where the border on the placeholder is covering the misaligned upload button, what if I remove the box-sizing?

I think in general it is better to have more predictable sizing to prevent janky layout shifts and the like. Quickly testing with TT4, if you set a border on the site logo block type, the placeholder will exceed its expected width and start to overlap the site title in the header part. 

<img width="302" alt="Screenshot 2024-08-08 at 8 24 35 PM" src="https://github.com/user-attachments/assets/2684246b-6c93-448d-b846-04bc9a8a3ace">

> So I am looking for a definite answer about whether the placeholder component should be styled by global styles or not.

This seems to be the crux of the matter and I don't know the answer, sorry 🙏 

At the risk of taking the easy way out, I'd like to defer to our design gurus on that. I suspect that either way, individual block instance styles and global styles should be rendered the same on placeholders.

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
> f you set a border on the site logo block type, the placeholder will exceed its expected width

I am not able to reproduce this. Did you do anything extra besides applying and building the PR?

carolinan commented:
With the option available in the toolbar, what if I remove the button that is in the placeholder?

carolinan committed:
Merge branch 'trunk' into add/site-logo-border

carolinan commented:
I am not able to unblock this.
